src/atomic/atomic_cell.rs

```diff
@@ -0,0 +1,695 @@
+
+            #[cfg(feature = "nightly")]
+            {
+                #[cfg(target_has_atomic = "8")]
+                atomic!(@check, $t, atomic::AtomicU8, $a, $atomic_op);
+                #[cfg(target_has_atomic = "16")]
+                atomic!(@check, $t, atomic::AtomicU16, $a, $atomic_op);
+                #[cfg(target_has_atomic = "32")]
+                atomic!(@check, $t, atomic::AtomicU32, $a, $atomic_op);
+                #[cfg(target_has_atomic = "64")]
+                atomic!(@check, $t, atomic::AtomicU64, $a, $atomic_op);
+            }
+
+            break $fallback_op
+        }
+    };
+}
+
+/// Returns `true` if operations on `AtomicCell<T>` are lock-free.
+fn atomic_is_lock_free<T>() -> bool {
+    atomic! { T, _a, true, false }
+}
+
+/// Atomically reads data from `src`.
+///
+/// This operation is sequentially consistent. If possible, an atomic instructions is used, and a
+/// global lock otherwise.
+unsafe fn atomic_load<T>(src: *mut T) -> T
+where
+    T: Copy,
+{
+    atomic! {
+        T, a,
+        {
+            a = &*(src as *const _ as *const _);
+            mem::transmute_copy(&a.load(Ordering::SeqCst))
+        },
+        {
+            let _lock = lock(src as usize);
+            ptr::read(src)
+        }
+    }
+}
+
+/// Atomically writes `val` to `dst`.
+///
+/// This operation is sequentially consistent. If possible, an atomic instructions is used, and a
+/// global lock otherwise.
+unsafe fn atomic_store<T>(dst: *mut T, val: T) {
+    atomic! {
+        T, a,
+        {
+            a = &*(dst as *const _ as *const _);
+            let res = a.store(mem::transmute_copy(&val), Ordering::SeqCst);
+            mem::forget(val);
+            res
+        },
+        {
+            let _lock = lock(dst as usize);
+            ptr::write(dst, val)
+        }
+    }
+}
+
+/// Atomically swaps data at `dst` with `val`.
+///
+/// This operation is sequentially consistent. If possible, an atomic instructions is used, and a
+/// global lock otherwise.
+unsafe fn atomic_swap<T>(dst: *mut T, val: T) -> T {
+    atomic! {
+        T, a,
+        {
+            a = &*(dst as *const _ as *const _);
+            let res = mem::transmute_copy(&a.swap(mem::transmute_copy(&val), Ordering::SeqCst));
+            mem::forget(val);
+            res
+        },
+        {
+            let _lock = lock(dst as usize);
+            ptr::replace(dst, val)
+        }
+    }
+}
+
+/// Atomically compares data at `dst` to `current` and, if equal byte-for-byte, swaps data at `dst`
+/// with `new`.
+///
+/// This operation is sequentially consistent. If possible, an atomic instructions is used, and a
+/// global lock otherwise.
+unsafe fn atomic_compare_and_swap<T>(dst: *mut T, current: T, new: T) -> T
+where
+    T: Copy,
+{
+    atomic! {
+        T, a,
+        {
+            a = &*(dst as *const _ as *const _);
+            mem::transmute_copy(
+                &a.compare_and_swap(
```

Comment:
I would recommend using the explicit `compare_exchange` or `compare_exchange_weak` operations rather than `compare_and_swap`. They return a `Result` which avoids the need to do a `byte_eq` afterwards.